src/sse.ts

```diff
@@ -0,0 +1,226 @@
+import fetch from "cross-fetch";
+import ReconnectingEventSource from "reconnecting-eventsource";
+
+import { ABLY_REALTIME_HOST, ABLY_REST_HOST } from "./config";
+
+interface AblyTokenDetails {
+  token: string;
+}
+
+interface AblyTokenRequest {
+  keyName: string;
+}
+
+export class AblySSEChannel {
+  private eventSource: ReconnectingEventSource | null = null;
+  private retryInterval: ReturnType<typeof setInterval> | null = null;
+  private debug: boolean;
+
+  constructor(
+    private userId: string,
+    private channel: string,
+    private ablyAuthUrl: string,
+    private messageHandler: (message: any) => void,
+    options?: {
+      debug?: boolean;
+    },
+  ) {
+    this.debug = options?.debug ?? false;
+  }
+
+  private log(message: string, ...args: any[]) {
+    if (this.debug) {
+      console.log("[SSE]", message, ...args);
+    }
+  }
+
+  private warn(message: string, ...args: any[]) {
+    console.warn("[SSE]", message, ...args);
+  }
+
+  private err(message: string, ...args: any[]): never {
+    if (this.debug) {
+      console.error("[SSE]", message, ...args);
+    }
+    throw new Error(message);
+  }
+
+  private async refreshTokenRequest() {
+    const res = await fetch(
+      `${this.ablyAuthUrl}?userId=${encodeURIComponent(this.userId)}`,
+      {
+        method: "get",
+        headers: {
+          "Content-Type": "application/json",
+        },
+      },
+    );
+
+    if (res.ok) {
+      const body = await res.json();
+      if (body.success) {
+        delete body.success;
+        const tokenRequest: AblyTokenRequest = body;
+
+        this.log("obtained new token request", tokenRequest);
+        return tokenRequest;
+      }
+    }
+
+    this.err("server did not release a token request", res);
+  }
+
+  private async refreshToken() {
+    const tokenRequest = await this.refreshTokenRequest();
+
+    const res = await fetch(
+      `${ABLY_REST_HOST}/keys/${encodeURIComponent(
```

Comment:
You can make this overridable with options and already set the default host to the bucket CNAME.

Reply:
Initially, I made them configurable, but then I removed the options since Ably does not actually have any other endpoints per environment so we will never really override these. So making them overridable is not really useful. Also, this client is Ably-specific so we would need to create a new client for any other back-end system we will have in the future.

We don't yet know the CNAME, still waiting for clarifications, so will only release the new SDK with those new CNAMEs plugged in and that will be it :)